src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,147 @@
+import mezz.jei.api.recipe.RecipeType;
+import mezz.jei.api.recipe.category.IRecipeCategory;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.phys.Vec2;
+import slimeknights.tconstruct.TConstruct;
+import slimeknights.tconstruct.library.client.GuiUtil;
+import slimeknights.tconstruct.library.materials.MaterialRegistry;
+import slimeknights.tconstruct.library.recipe.tinkerstation.building.ToolBuildingRecipe;
+import slimeknights.tconstruct.library.tools.helper.ToolBuildHandler;
+import slimeknights.tconstruct.library.tools.helper.TooltipUtil;
+import slimeknights.tconstruct.library.tools.item.IModifiableDisplay;
+import slimeknights.tconstruct.library.tools.layout.LayoutSlot;
+import slimeknights.tconstruct.library.tools.part.IToolPart;
+import slimeknights.tconstruct.tools.TinkerTools;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+public class ToolBuildingCategory implements IRecipeCategory<ToolBuildingRecipe> {
+  private static final ResourceLocation BACKGROUND_LOC = TConstruct.getResource("textures/gui/jei/tinker_station.png");
+  private static final Component TITLE = TConstruct.makeTranslation("jei", "tinkering.tool_building");
+  @Getter
+  private final IDrawable icon;
+  @Getter
+  private final IDrawable background;
+  private final IDrawable anvil, slot;
+  private static final int WIDTH = 128;
+  private static final int HEIGHT = 66;
+  private static final int SLOT_SIZE = 18;
+  private static final int ITEM_SIZE = 16;
+
+  public ToolBuildingCategory(IGuiHelper guiHelper) {
+    this.icon = guiHelper.createDrawableItemStack(TinkerTools.pickaxe.get().getRenderTool());
+    this.background = guiHelper.createDrawable(BACKGROUND_LOC, 128, 77, WIDTH, HEIGHT);
+    this.slot = guiHelper.createDrawable(BACKGROUND_LOC, 24, 14, SLOT_SIZE, SLOT_SIZE);
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<IToolPart> parts = recipe.getToolParts();
+    List<ItemStack> extraItems = recipe.getIngredients().stream().flatMap(ingredient -> Arrays.stream(ingredient.getItems())).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    // every part and its material variants
+    List<ItemStack> allInputParts = parts.stream()
+        .flatMap(part -> MaterialRegistry.getInstance().getVisibleMaterials().stream()
+          .filter(part::canUseMaterial)
+          .flatMap(mat -> Stream.of(part.withMaterial(mat.getIdentifier()))))
+        .toList();
+
+    if (parts.size() + extraItems.size() != layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count and layout slot count for %s do not match!", recipe.getId()));
+      return;
```

Comment:
Is it possible for us to recover instead of giving up? That is:

* If the layout has too many slots, just display the extra slots as empty.
* If the layout has too few slots, add the extra slots in some hardcoded location.

don't put too much effort into this, just want to give the user as much usable info as possible; it will mostly end up helping modders I think.